### PR TITLE
[Android R] Sync keyboard animation with view insets vs Android 11/R/API 30 WindowInsetsAnimation

### PR DESCRIPTION
**Pull request by GaryQian**
Adds a `WindowInsetsAnimation.Callback` and `OnApplyWindowInsetsListener` class that handles sending view inset changes from the keyboard animating to the framework.

When keyboard is shown or hidden, a call to `onApplyWindowInsets()` is made immediately with the keyboard's final state. This causes a flicker as the animation kicks in a frame later. To solve this, we check for this initial call and capture the state, deferring it being called until the animation is completed.

During the animation, we pass the window insets to the framework on each frame.

There is an issue where the `onProgress` WindowInsets received has a bottom ime inset that is higher than expected. It is actually the sum of the navigation bars on top of the keyboard height. There is a workaround implemented to merge the navbar height with the keyboard height to resolve this.

Partially resolves https://github.com/flutter/flutter/issues/62876

### CODE REVIEW

File: shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java [hunk 1]
```diff
@@ -134,6 +160,117 @@ public void sendAppPrivateCommand(String action, Bundle data) {
...
+        // Store the view and insets for us in onEnd() below
+        lastWindowInsets = windowInsets;
+      }
+      if (deferredInsets) {
+        return WindowInsets.CONSUMED;
+      }
+      return view.onApplyWindowInsets(windowInsets);
+    }
+
+    @TargetApi(30)
```
Comment (xster) [inline shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java]: Do you need this? This whole class is api 30 no?

File: shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java [hunk 2]
```diff
@@ -65,6 +73,24 @@ public TextInputPlugin(
       afm = null;
     }
 
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
+      int mask = 0;
+      if ((View.SYSTEM_UI_FLAG_HIDE_NAVIGATION & mView.getWindowSystemUiVisibility()) == 0) {
+        mask = mask | WindowInsets.Type.navigationBars();
+      }
+      if ((View.SYSTEM_UI_FLAG_FULLSCREEN & mView.getWindowSystemUiVisibility()) == 0) {
```
Comment (xster) [inline shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java:81]: does full screen also independently affect navigation bar above too?

File: shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java [hunk 3]
```diff
@@ -65,6 +73,24 @@ public TextInputPlugin(
...
+        mask = mask | WindowInsets.Type.statusBars();
+      }
+      imeSyncCallback =
+          new ImeSyncDeferringInsetsCallback(
+              view,
+              mask, // Overlay
+              WindowInsets.Type.ime() // Deferred
+              );
+      view.setWindowInsetsAnimationCallback(imeSyncCallback);
```
Comment (xster) [inline shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java:90]: add comment on why you need 2

File: shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java [hunk 4]
```diff
@@ -65,6 +73,24 @@ public TextInputPlugin(
...
+      view.setOnApplyWindowInsetsListener(imeSyncCallback);
```
Comment (xster) [inline shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java:91]: do you remove this in `destroy`?

File: shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java [hunk 5]
```diff
@@ -134,6 +160,117 @@ public void sendAppPrivateCommand(String action, Bundle data) {
     restartAlwaysRequired = isRestartAlwaysRequired();
   }
 
+  // Loosely based off of
+  // https://github.com/android/user-interface-samples/blob/master/WindowInsetsAnimation/app/src/main/java/com/google/android/samples/insetsanimation/RootViewDeferringInsetsCallback.kt
+  // When the IME is shown or hidden, it sends an onApplyWindowInsets call with the
+  // final state of the IME. This defers the final call to allow the animation to
```
Comment (xster) [inline shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java:166]: Sentence not clear to me. What's 'this'? 'final call' to what?

File: shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java [hunk 6]
```diff
@@ -65,6 +73,24 @@ public TextInputPlugin(
...
+      if ((View.SYSTEM_UI_FLAG_HIDE_NAVIGATION & mView.getWindowSystemUiVisibility()) == 0) {
+        mask = mask | WindowInsets.Type.navigationBars();
+      }
+      if ((View.SYSTEM_UI_FLAG_FULLSCREEN & mView.getWindowSystemUiVisibility()) == 0) {
+        mask = mask | WindowInsets.Type.statusBars();
+      }
+      imeSyncCallback =
+          new ImeSyncDeferringInsetsCallback(
+              view,
+              mask, // Overlay
```
Comment (xster) [inline shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java:87]: oh I see what this is after reading your other class constructor. Add comments to make this very clear. 

e.g. these are the things that'll reflect in flutter right away

these are the things that will animate

File: shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java [hunk 7]
```diff
@@ -134,6 +160,117 @@ public void sendAppPrivateCommand(String action, Bundle data) {
...
+    private View view;
+    private WindowInsets lastWindowInsets;
+    private boolean deferredInsets = false;
+    private boolean started = false;
+
+    ImeSyncDeferringInsetsCallback(View view, int persistentInsetTypes, int deferredInsetTypes) {
+      super(WindowInsetsAnimation.Callback.DISPATCH_MODE_CONTINUE_ON_SUBTREE);
+      this.overlayInsetTypes = overlayInsetTypes;
+      this.deferredInsetTypes = deferredInsetTypes;
+      this.view = view;
```
Comment (xster) [inline shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java:186]: preconditions checknotnull

File: shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java [hunk 8]
```diff
@@ -134,6 +160,117 @@ public void sendAppPrivateCommand(String action, Bundle data) {
...
+    }
+
+    @Override
+    public WindowInsets onApplyWindowInsets(View view, WindowInsets windowInsets) {
+      this.view = view;
+      if (!started) {
```
Comment (xster) [inline shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java:192]: ya, these apis are pretty confusing. I think this needs a lot more comments. Comments at the class level for what the flow should be for the persistent vs delayed stuff are. When this !started should happen, when should deferredInsets happen.

i.e. I don't think the next maintainer would know what to do here if there was a bug and you're out.

File: shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java [hunk 9]
```diff
@@ -134,6 +160,117 @@ public void sendAppPrivateCommand(String action, Bundle data) {
...
+      if ((animation.getTypeMask() & deferredInsetTypes) != 0) {
+        deferredInsets = true;
+      }
+    }
+
+    @TargetApi(30)
+    @Override
+    public WindowInsetsAnimation.Bounds onStart(
+        WindowInsetsAnimation animation, WindowInsetsAnimation.Bounds bounds) {
+      if (deferredInsets && (animation.getTypeMask() & deferredInsetTypes) != 0) {
```
Comment (xster) [inline shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java:212]: comments here too. What's the difference between onprepare and onstart? i.e. why not just set both deferredInsets and started to true here? Can you start and not prepare?

File: shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java [hunk 10]
```diff
@@ -134,6 +160,117 @@ public void sendAppPrivateCommand(String action, Bundle data) {
...
+          Insets.of(
+              0,
+              0,
+              0,
+              Math.max(
+                  insets.getInsets(deferredInsetTypes).bottom
+                      - insets.getInsets(overlayInsetTypes).bottom,
+                  0));
+      builder.setInsets(deferredInsetTypes, newImeInsets);
+      // Directly call onApplyWindowInsets as we want to skip this class' version of this call.
```
Comment (xster) [inline shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java:246]: This isn't super clear. Maybe elaborate a bit more.

File: shell/platform/android/test/io/flutter/plugin/editing/TextInputPluginTest.java [hunk 11]
```diff
@@ -632,6 +637,96 @@ public void sendAppPrivateCommand_hasData() throws JSONException {
     assertEquals("actionData", bundleCaptor.getValue().getCharSequence("data"));
   }
 
+  @Test
+  @TargetApi(30)
+  @Config(sdk = 30)
+  public void ime_windowInsetsSync() {
+    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.R) {
+      return;
```
Comment (xster) [inline shell/platform/android/test/io/flutter/plugin/editing/TextInputPluginTest.java:645]: if you say this needs sdk 30 and it's not sdk 30, we should just crash :D

File: shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java (hunk 8 above)
Comment (GaryQian) [reply shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java:192]: I'll make a pass and document the whole thing

File: shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java (hunk 9 above)
Comment (GaryQian) [reply shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java]: I initially thought that both were necessary, turns out only start is required.

File: shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java (hunk 2 above)
Comment (GaryQian) [reply shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java:84]: These are the two values we use to compute viewPadding, so for consistency, I include it together to keep in sync with FlutterView.onApplyWindowInsets. Any padding we consume will also be accounted for here this way.

File: shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java [hunk 12]
```diff
@@ -134,6 +163,139 @@ public void sendAppPrivateCommand(String action, Bundle data) {
     restartAlwaysRequired = isRestartAlwaysRequired();
   }
 
+  // Loosely based off of
+  // https://github.com/android/user-interface-samples/blob/master/WindowInsetsAnimation/app/src/main/java/com/google/android/samples/insetsanimation/RootViewDeferringInsetsCallback.kt
+  //
+  // When the IME is shown or hidden, it immediately sends an onApplyWindowInsets call
```
Comment (xster) [inline shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java:169]: Ah, this is much better, thanks! Does this class also handle all the non-animating inset changes? Like the nav bar, status bar going in and out of fullscreen etc? How do all the phases like onStart, onApplyWindowInsets etc affect those insets?

File: shell/platform/android/test/io/flutter/plugin/editing/TextInputPluginTest.java [hunk 13]
```diff
@@ -632,6 +644,92 @@ public void sendAppPrivateCommand_hasData() throws JSONException {
...
+    WindowInsets imeInsets1 = builder.build();
+
+    builder.setInsets(WindowInsets.Type.ime(), Insets.of(0, 0, 0, 200));
+    builder.setInsets(WindowInsets.Type.navigationBars(), Insets.of(10, 10, 10, 0));
+    WindowInsets deferredInsets = builder.build();
+
+    ArgumentCaptor<FlutterRenderer.ViewportMetrics> viewportMetricsCaptor =
+        ArgumentCaptor.forClass(FlutterRenderer.ViewportMetrics.class);
+
+    imeSyncCallback.onApplyWindowInsets(testView, deferredInsets);
```
Comment (xster) [inline shell/platform/android/test/io/flutter/plugin/editing/TextInputPluginTest.java:687]: I was gonna say I'm surprised you have to do this on your callback directly but then I actually can't find this on any of the robolectric shadows https://github.com/robolectric/robolectric/search?q=onApplyWindowInsets 🤷‍♂️

**Review by xster — COMMENTED** [review]
LGTM

File: shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java (hunk 12 above)
Comment (GaryQian) [reply shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java:169]: Non-animating changes are just passed right on through to FlutterView.onApplyWindowInsets without impact. The ime-type filtering also prevents this from doing anything on any non-ime animations. During animations, changes to the window are handled by the animation controller, which will end the current animation, and restart a new one with the modified settings to continue. For example, rotating the device mid-keyboard animation just causes the animation to continue where it left off but scaled to the rotated window.

File: shell/platform/android/test/io/flutter/plugin/editing/TextInputPluginTest.java (hunk 13 above)
Comment (GaryQian) [reply shell/platform/android/test/io/flutter/plugin/editing/TextInputPluginTest.java:687]: 🤷‍♂️

**Comment by xster** [comment]
🙌

**Pull request merged**